lib/Runtime/Library/JSONStringifier.cpp

```diff
@@ -0,0 +1,817 @@
+//-------------------------------------------------------------------------------------------------------
+// Copyright (C) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE.txt file in the project root for full license information.
+//-------------------------------------------------------------------------------------------------------
+
+#include "RuntimeLibraryPch.h"
+
+namespace Js
+{
+
+JSONStringifier::JSONStringifier(_In_ ScriptContext* scriptContext) :
+    scriptContext(scriptContext),
+    replacerFunction(nullptr),
+    totalStringLength(0),
+    indentLength(0),
+    gapLength(0),
+    gap(nullptr),
+    propertyList(nullptr)
+{
+}
+
+void
+JSONStringifier::SetStringGap(_In_ JavascriptString* spaceString)
+{
+    this->gapLength = min(MaxGapLength, spaceString->GetLength());
+    if (this->gapLength != 0)
+    {
+        this->gap = RecyclerNewArrayLeaf(this->scriptContext->GetRecycler(), char16, this->gapLength);
+        wmemcpy_s(gap, this->gapLength, spaceString->GetString(), this->gapLength);
+    }
+}
+
+void
+JSONStringifier::SetNumericGap(charcount_t spaceCount)
+{
+    this->gapLength = spaceCount;
+    if (this->gapLength != 0)
+    {
+        this->gap = RecyclerNewArrayLeaf(this->scriptContext->GetRecycler(), char16, this->gapLength);
+        wmemset(gap, _u(' '), this->gapLength);
+    }
+}
+
+void
+JSONStringifier::ReadSpace(_In_opt_ Var space)
+{
+    if (space != nullptr)
+    {
+        switch (JavascriptOperators::GetTypeId(space))
+        {
+        case TypeIds_Integer:
+            this->SetNumericGap(static_cast<charcount_t>(max(0, min(static_cast<int>(MaxGapLength), TaggedInt::ToInt32(space)))));
+
+            break;
+        case TypeIds_Number:
+        case TypeIds_NumberObject:
+        case TypeIds_Int64Number:
+        case TypeIds_UInt64Number:
+        {
+            double numericSpace = JavascriptConversion::ToInteger(space, this->scriptContext);
+            if (numericSpace > 0)
+            {
+                SetNumericGap(numericSpace < static_cast<double>(MaxGapLength) ? static_cast<charcount_t>(numericSpace) : MaxGapLength);
+            }
+            break;
+        }
+        case TypeIds_String:
+            this->SetStringGap(JavascriptString::FromVar(space));
+            break;
+        case TypeIds_StringObject:
+            this->SetStringGap(JavascriptConversion::ToString(space, this->scriptContext));
```

Comment:
Is this correct? Some experimentation seems to suggest that non-string non-number values here have no effect.

Reply:
Assuming this corresponds to the third parameter of `JSON.stringify(obj, replacer, space)`

Reply:
Oh nevermind; I misread this as the case for `Object`, not `StringObject`.